lang/c/src/io.c

```diff
@@ -345,7 +345,7 @@ avro_write_file(struct _avro_writer_file_t *writer, void *buf, int64_t len)
 	if (len > 0) {
 		rval = fwrite(buf, len, 1, writer->fp);
 		if (rval == 0) {
-			return feof(writer->fp) ? EOF : 0;
+			return feof(writer->fp) ? EOF : (errno ? errno : EIO);
```

Comment:
The call to feof() does not seem to have much point, except if there is also a call to ferror() as well.
Returning errno is probably not a great idea either, as possibly not set.  This whole line should probably just return unconditionally EIO.

Digging a bit deeper into the C11 standard and glibc implementation, after a failed fwrite, the file position is indeterminate.  Further writes will result in a corrupt file.  If the file is closed as it is, the last record will be truncated.